crypto/bls12/bls12.go

```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
 	g2 := bls12.NewG2()
-	sig := bls12.PointG2{}
-	var participants crypto.Bitfield
-	for id, s := range signatures {
-		g2.Add(&sig, &sig, s.s)
-		participants.Add(id)
-	}
-	return &AggregateSignature{sig: sig, participants: participants}
+	g2.MulScalarBig(&p, point, curveOrder)
+	return g2.IsZero(&p)
 }
 
-// Verify verifies a signature given a hash.
-func (bc *bls12Crypto) Verify(sig consensus.Signature, hash consensus.Hash) bool {
-	s := sig.(*Signature)
-	replica, ok := bc.mods.Configuration().Replica(sig.Signer())
-	if !ok {
-		bc.mods.Logger().Infof("bls12Crypto: got signature from replica whose ID (%d) was not in the config", sig.Signer())
-	}
-	pk := replica.PublicKey().(*PublicKey)
-	p, err := bls12.NewG2().HashToCurve(hash[:], domain)
+func (bls *bls12Base) coreSign(message []byte, domainTag []byte) (*bls12.PointG2, error) {
+	pk := bls.getPrivateKey()
+	g2 := bls12.NewG2()
+	point, err := g2.HashToCurve(message, domainTag)
 	if err != nil {
-		return false
+		return nil, err
 	}
-	engine := bls12.NewEngine()
-	engine.AddPairInv(&bls12.G1One, s.s)
-	engine.AddPair(pk.p, p)
-	return engine.Result().IsOne()
+	// multiply the point by the secret key, storing the result in the same point variable
+	g2.MulScalarBig(point, point, pk.p)
+	return point, nil
 }
 
-// VerifyAggregateSignature verifies an aggregated signature.
-// It does not check whether the aggregated signature contains a quorum of signatures.
-func (bc *bls12Crypto) VerifyAggregateSignature(agg consensus.ThresholdSignature, hash consensus.Hash) bool {
-	sig, ok := agg.(*AggregateSignature)
-	if !ok {
+func (bls *bls12Base) coreVerify(pubKey *PublicKey, message []byte, signature *bls12.PointG2, domainTag []byte) bool {
+	if !bls.subgroupCheck(signature) {
 		return false
 	}
-	pubKeys := make([]*PublicKey, 0)
-	sig.participants.ForEach(func(id hotstuff.ID) {
-		replica, ok := bc.mods.Configuration().Replica(id)
-		if !ok {
-			return
-		}
-		pubKeys = append(pubKeys, replica.PublicKey().(*PublicKey))
-	})
-	ps, err := bls12.NewG2().HashToCurve(hash[:], domain)
+	g2 := bls12.NewG2()
+	messagePoint, err := g2.HashToCurve(message, domainTag)
 	if err != nil {
-		bc.mods.Logger().Error(err)
 		return false
 	}
 	engine := bls12.NewEngine()
-	engine.AddPairInv(&bls12.G1One, &sig.sig)
-	for _, pub := range pubKeys {
-		engine.AddPair(pub.p, ps)
-	}
+	engine.AddPairInv(&bls12.G1One, signature)
+	engine.AddPair(pubKey.p, messagePoint)
 	return engine.Result().IsOne()
 }
 
-// TODO: I'm not sure to what extent we are vulnerable to a rogue public key attack here.
-// As far as I can tell, this is not a problem right now because we do not yet support reconfiguration,
-// and all public keys are known by all replicas.
+func (bls *bls12Base) popProve() *bls12.PointG2 {
+	pubKey := bls.getPrivateKey().Public().(*PublicKey)
+	proof, err := bls.coreSign(pubKey.ToBytes(), domainPOP)
+	if err != nil {
+		bls.mods.Logger().Panicf("failed to generate proof of possession: %v", err)
```

Comment:
Should logged messages be capitalized? `"Failed to ..."`

Reply:
I assume this will panic. What's the policy for logging a panic vs emitting it to stdout; I assume this is the difference... There are other places where regular panic is used.

Reply:
Actually, I don't have a policy for that.

Here are some more differences when compared with `panic`:

`logging.Panic` and `logging.Panicf`:
- can only be used where the logger module is accessible
- the message will be logged even if the panic is recovered
- `Panicf` supports string formatting